src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2959,6 +2932,139 @@ def add(attr):
                 numdisks = num
         return '%dK' % 2 ** ((numdisks * 4) - 1).bit_length()
 
+    @item_method
+    @accepts(Str('id', required=True))
+    async def attachments(self, oid):
+        """
+        Return a list of services dependent of this dataset.
+
+        Responsible for telling the user whether there is a related
+        share, asking for confirmation.
+
+        Example return value:
+        [
+          {
+            "type": "NFS Share",
+            "service": "nfs",
+            "attachments": ["/mnt/tank/work"]
+          }
+        ]
+        """
+        result = []
+        dataset = await self._get_instance(oid)
+        path = self.__attachments_path(dataset)
+        if path:
+            for delegate in self.attachment_delegates:
+                attachments = {"type": delegate.title, "service": delegate.service, "attachments": []}
+                for attachment in await delegate.query(path, True):
+                    attachments["attachments"].append(await delegate.get_attachment_name(attachment))
+                if attachments["attachments"]:
+                    result.append(attachments)
+        return result
+
+    def __attachments_path(self, dataset):
+        if dataset['type'] == 'FILESYSTEM':
+            return dataset['mountpoint']
+
+        if dataset['type'] == 'VOLUME':
+            return os.path.join('/mnt', dataset['name'])
+
+    @item_method
+    @accepts(Str('id', required=True))
+    async def processes(self, oid):
+        """
+        Return a list of processes using this dataset.
+
+        Example return value:
+
+        [
+          {
+            "pid": 2520,
+            "name": "smbd",
+            "service": "cifs"
+          },
+          {
+            "pid": 97778,
+            "name": "minio",
+            "cmdline": "/usr/local/bin/minio -C /usr/local/etc/minio server --address=0.0.0.0:9000 --quiet /mnt/tank/wk"
+          }
+        ]
+        """
+        result = []
+        dataset = await self._get_instance(oid)
+        path = self.__attachments_path(dataset)
+        if path:
+            lsof = await run('lsof',
```

Comment:
I am not sure how practical this is. I have tried this in a pool with 100 datasets, lsof takes over 2 minutes to run and uses a lot of CPU.
Can we investigate some way to make it faster and/or alternatives?